cmd/sourced/cmd/web.go

```diff
@@ -75,11 +55,117 @@ func openUI() error {
 	return nil
 }
 
+func checkFailFast(stdout *bytes.Buffer) (bool, error) {
+	err := compose.RunWithIO(context.Background(),
+		os.Stdin, stdout, nil, "port", containerName, "8088")
+	if workdir.ErrMalformed.Is(err) || dir.ErrNotExist.Is(err) || dir.ErrNotValid.Is(err) {
+		return true, err
+	}
+
+	if err != nil {
+		return false, err
+	}
+
+	return false, nil
+}
+
+func waitForContainer(stdout *bytes.Buffer) {
+	for {
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, stdout, nil, "port", containerName, "8088"); err == nil {
+			break
+		}
+
+		time.Sleep(1 * time.Second)
+	}
+}
+
+var newLineFormatter = regexp.MustCompile(`(\r\n|\r|\n)`)
+
+func normalizeNewLine(s string) string {
+	return newLineFormatter.ReplaceAllString(s, "\n")
+}
+
+func runMonitor(ch chan<- error) {
+	go func() {
+		client, err := docker.GetClient()
+		if err != nil {
+			ch <- errors.Wrap(err, "cannot get docker client")
+			return
+		}
+
+		var stdout bytes.Buffer
+		if err := compose.RunWithIO(context.Background(),
+			os.Stdin, &stdout, nil, "ps", "-q"); err != nil {
+			ch <- errors.Wrap(err, "cannot get services' container id")
+			return
+		}
+
+		containerIds := strings.Split(
+			normalizeNewLine(strings.TrimSpace(stdout.String())),
+			"\n",
+		)
+
+		for {
+			for _, contID := range containerIds {
+				cont, err := docker.InfoWithClient(client, contID)
```

Comment:
why do you need calling docker directly if `docker-compose ps` which you run above already returns you all the statuses?

Reply:
the reason is that I'd have to parse the stdout and I found it not much reliable, that's why I decided to use the client instead.